c++/src/capnp/cc_capnp_library.bzl

```diff
@@ -49,8 +49,18 @@ def _capnp_gen_impl(ctx):
 
     args.add_all([s for s in ctx.files.srcs])
 
+    # process language extras, starting with files we should include
+    extra_inputs = ctx.files.language_extras or []
+
+    # then compute input flags, taking the parent of the first file in each extra as the include path
+    if len(ctx.attr.language_extras) > 0:
+        for extra in ctx.attr.language_extras:
+            files = extra.files.to_list()
+            if len(files) > 0:
+                args.add_all(["-I", files[0].dirname.removesuffix("/capnp")])
+
     ctx.actions.run(
-        inputs = inputs + ctx.files._capnpc_cxx + ctx.files._capnpc_capnp + ctx.files._capnp_system,
+        inputs = inputs + ctx.files._capnpc_cxx + ctx.files._capnpc_capnp + ctx.files._capnp_system + extra_inputs,
```

Comment:
@mikea Here's the entire change. If the user passes additional targets via the `language_extras` attribute, they are included inputs in the run, and each directory _containing_ the language extras is included via `-I`.

The effect is that the user can pass arbitrary dependencies to be exposed in the sandbox, and can include them easily within the `capnp` sources.

We're using this change downstream in an internal Bazel project, and it's working great, specifically to generate Java capnp sources.

I actually know exactly what you're talking about re/`rules_proto`. It is a little brain-melting to get aspects all working together properly, certainly not for the faint of heart. I've died on that same hill before.

So, I think you're right, it would be _fantastic_ to get true first-class Bazel support for capnp in the same way Protocol Buffers is supported, including easy generation of different languages. But in the meantime, this gives capnp users an escape hatch in Bazel builds where there is no explicit support for their language yet.

Thus, these attributes could be kept to use in-house language implementations, or to facilitate easy testing, etc etc, alongside future attributes or rules or aspects which bring support to parity.